src/backend/vulkan/src/window.rs

```diff
@@ -435,6 +469,125 @@ impl hal::Surface<Backend> for Surface {
     }
 }
 
+#[derive(Debug)]
+pub struct SurfaceImage {
+    pub(crate) index: hal::SwapImageIndex,
+    views: Arc<Vec<native::ImageView>>,
+}
+
+impl Borrow<native::ImageView> for SurfaceImage {
+    fn borrow(&self) -> &native::ImageView {
+        &self.views[self.index as usize]
+    }
+}
+
+impl hal::PresentationSurface<Backend> for Surface {
+    type SwapchainImage = SurfaceImage;
+
+    /// Set up the swapchain associated with the surface to have the given format.
+    unsafe fn configure_swapchain(
+        &mut self, device: &Device, config: hal::SwapchainConfig
+    ) -> Result<(), hal::window::CreationError> {
+        use ash::version::DeviceV1_0;
+        use hal::Device;
+
+        let format = config.format;
+        let old = match self.swapchain.take() {
+            Some(ssc) => {
+                ssc.device.0.destroy_fence(ssc.fence.0, None);
+                self.stale_views.push((ssc.device, ssc.semaphore, ssc.views));
+                Some(ssc.swapchain)
+            }
+            None => None,
+        };
+
+        let (swapchain, images) = device.create_swapchain(self, config, old)?;
```

Comment:
There's a validation error on program exit about the swapchain not being destroyed prior to the surface being destroyed. The swapchain ought to be destroyed when the surface is destroyed.